feat(typeform): add v2 detector for new key formats

ggmolly opened the pull request:
### Description:
Typeform has apparently introduced a new format for API keys, according to the [API documentation](https://www.typeform.com/developers/get-started/personal-access-token/) and a key I have generated, I wrote a second version of the Typeform.

### Checklist:
* [x] Tests passing (`make test-community`)?
* [x] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/welcome/install/#local-installation))?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/trufflesecurity/trufflehog?pullRequest=3660) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/trufflesecurity/trufflehog?pullRequest=3660) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/trufflesecurity/trufflehog?pullRequest=3660) it.</sub>

rgmz commented on pkg/detectors/typeform/v2/typeform.go:
```diff
@@ -0,0 +1,96 @@
+	client = common.SaneHttpClient()
+
+	// Make sure that your group is surrounded in boundary characters such as below to reduce false positives.
+	keyPat = regexp.MustCompile(`\btfp_[a-zA-Z0-9_\.=\-]{40,59}\b`)
```
The comment's suggestion to use a boundary character is actually bad advice, it won't match if the last character is `.=-`.

One possible alternative is to say "anything but":
```suggestion
	keyPat = regexp.MustCompile(`\btfp_[a-zA-Z0-9_\.=\-]{40,59}(?:[^a-zA-Z0-9_\.=\-]|\z)`)
```

rgmz commented on pkg/detectors/typeform/v2/typeform.go:
```diff
@@ -0,0 +1,96 @@
+// Keywords are used for efficiently pre-filtering chunks.
+// Use identifiers in the secret preferably, or the provider name.
+func (s Scanner) Keywords() []string {
+	return []string{"typeform", "tfp_"}
```
```suggestion
	return []string{"tfp_"}
```

rgmz commented on pkg/detectors/typeform/v2/typeform.go:
```diff
@@ -0,0 +1,96 @@
+		}
+
+		if verify {
+			req, err := http.NewRequestWithContext(ctx, "GET", "https://api.typeform.com/me", nil)
```
The older convention for detector logic has a bunch of issues. I'd recommend moving this into a separate function (for the sake of `defer`) and having explicit error handling + status code checks.

e.g., https://github.com/trufflesecurity/trufflehog/blob/abaacd7e8af3019d622848641a095e471b6daf74/pkg/detectors/alchemy/alchemy.go#L68

ggmolly commented:
Upon further testing I wasn't able to neither generate nor find a key with any of these characters: `.=-`. Therefore, I removed them from the regex entirely. However, I refactored the verification part of the code, removed the extra keyword, and squashed the changes.

kashifkhan0771 commented on pkg/detectors/typeform/v2/typeform.go:
```diff
@@ -0,0 +1,107 @@
+
+	matches := keyPat.FindAllString(dataStr, -1)
+
+	if len(matches) == 0 {
```
Not required. If length is zero the loop will be skipped.

kashifkhan0771 commented on pkg/detectors/typeform/v2/typeform.go:
```diff
@@ -0,0 +1,107 @@
+	if err = json.NewDecoder(res.Body).Decode(&response); err != nil {
+		return false, response, err
+	}
+	if res.StatusCode >= 200 && res.StatusCode < 300 {
```
We should be more specific for success validation status codes.

kashifkhan0771 commented on pkg/detectors/typeform/v2/typeform.go:
```diff
@@ -0,0 +1,107 @@
+		return false, response, err
+	}
+	if res.StatusCode >= 200 && res.StatusCode < 300 {
+		return true, response, nil
```
Not handling the fail validation status like 401 or 403 etc? Error should be returned in case of unexpected status code.

kashifkhan0771 commented on pkg/detectors/typeform/v2/typeform_test.go:
```diff
@@ -0,0 +1,120 @@
+//go:build detectors
```
According to new structure the `test.go` file should contain pattern test cases and the `_integration_test.go` file should have the integration test cases.

ggmolly committed:
feat(typeform): add v2 detector for new key formats

docs:
- https://www.typeform.com/developers/get-started/personal-access-token/

40 to 59 chars is set because we've seen tokens with an extra 15 bytes at the end, the base key is 40 bytes long

ggmolly commented:
1. Updated the code to consider only HTTP 200 as valid response, other codes return an error
2. Removed the redundant check
3. Changed tests to match the new structure
(squashed again)

rosecodym commented:
Thank you for this!
